josh-proxy/src/lib.rs

```diff
@@ -282,6 +287,65 @@ pub fn process_repo_update(repo_update: RepoUpdate) -> josh::JoshResult<String>
     Ok("".to_string())
 }
 
+fn split_changes(
+    repo: &git2::Repository,
+    changes: &mut Vec<(String, git2::Oid, String)>,
+    base: git2::Oid,
+) -> josh::JoshResult<()> {
+    if base == git2::Oid::zero() {
+        return Ok(());
+    }
+
+    let commits: Vec<git2::Commit> = changes
+        .iter()
+        .map(|(_, commit, _)| repo.find_commit(*commit).unwrap())
+        .collect();
+
+    let mut trees: Vec<git2::Tree> = commits
+        .iter()
+        .map(|commit| commit.tree().unwrap())
+        .collect();
+
+    trees.insert(0, repo.find_commit(base)?.tree()?);
+
+    let diffs: Vec<git2::Diff> = (1..trees.len())
+        .map(|i| {
+            repo.diff_tree_to_tree(Some(&trees[i - 1]), Some(&trees[i]), None)
+                .unwrap()
+        })
+        .collect();
+
+    let mut moved = std::collections::HashSet::new();
+    let mut bases = vec![base];
+    for _ in 0..changes.len() {
+        let mut new_bases = vec![];
+        for base in bases.iter() {
+            for i in 0..diffs.len() {
+                if moved.contains(&i) {
+                    continue;
+                }
+                let diff = &diffs[i];
+                let parent = repo.find_commit(*base)?;
+                if let Ok(mut index) = repo.apply_to_tree(&parent.tree()?, &diff, None) {
+                    moved.insert(i);
+                    let new_tree = repo.find_tree(index.write_tree_to(repo)?)?;
+                    let new_commit = josh::history::rewrite_commit(
+                        repo,
+                        &repo.find_commit(changes[i].1)?,
+                        &vec![&parent],
+                        &new_tree,
+                    )?;
+                    changes[i].1 = new_commit;
+                    new_bases.push(new_commit);
+                }
+            }
+        }
```

Comment:
add break if moved.len == changes.len

Reply:
That has the potential for an infinite loop if there is a leftover change that cannot be rebased